runtime/contract_update_validation.go

```diff
@@ -0,0 +1,456 @@
+	newDeclaration ast.Declaration,
+) error {
+
+	// Do not allow converting between different types of composite declarations:
+	// e.g: - 'contracts' and 'contract-interfaces',
+	//      - 'structs' and 'enums'
+	if oldDeclaration.DeclarationKind() != newDeclaration.DeclarationKind() {
+		return &InvalidDeclarationKindChangeError{
+			name:    oldDeclaration.DeclarationIdentifier().Identifier,
+			oldKind: oldDeclaration.DeclarationKind().Name(),
+			newKind: newDeclaration.DeclarationKind().Name(),
+			Range:   ast.NewRangeFromPositioned(newDeclaration),
+		}
+	}
+
+	parentDecl := validator.currentDecl
+	validator.currentDecl = newDeclaration
+	defer func() {
+		validator.currentDecl = parentDecl
+	}()
+
+	// If the same decl is already visited, then do not check again.
+	// This also avoids getting stuck on circular dependencies between composite decls.
+	if validator.visited[newDeclaration] {
+		return nil
+	}
+	validator.visited[newDeclaration] = true
+
+	err := validator.checkFields(oldDeclaration, newDeclaration)
+	if err != nil {
+		return err
+	}
+
+	err = validator.checkNestedDeclarations(oldDeclaration, newDeclaration)
+	if err != nil {
+		return err
+	}
+
+	switch newDecl := newDeclaration.(type) {
+	case *ast.CompositeDeclaration:
+		oldDecl := oldDeclaration.(*ast.CompositeDeclaration)
+		return validator.checkConformances(oldDecl, newDecl)
+	}
+
+	return nil
+}
+
+func (validator *ContractUpdateValidator) checkFields(
+	oldDeclaration ast.Declaration,
+	newDeclaration ast.Declaration,
+) error {
+
+	oldFields := oldDeclaration.DeclarationMembers().FieldsByIdentifier()
+	newFields := newDeclaration.DeclarationMembers().Fields()
+
+	// Updated contract has to have at-most the same number of field as the old contract.
+	// Any additional field may cause crashes/garbage-values when deserializing the
+	// already-stored data. However, having less number of fields is fine for now. It will
+	// only leave some unused-data, and will not do any harm to the programs that are running.
+
+	// This is a fail-fast check.
+	if len(oldFields) < len(newFields) {
+		currentDeclName := newDeclaration.DeclarationIdentifier()
+		return &TooManyFieldsError{
+			declName:       currentDeclName.Identifier,
+			expectedFields: len(oldFields),
+			foundFields:    len(newFields),
+			Range:          ast.NewRangeFromPositioned(currentDeclName),
+		}
+	}
+
+	for _, newField := range newFields {
+		oldField := oldFields[newField.Identifier.Identifier]
+		if oldField == nil {
+			return &ExtraneousFieldError{
+				declName:  newDeclaration.DeclarationIdentifier().Identifier,
+				fieldName: newField.Identifier.Identifier,
+				Range:     ast.NewRangeFromPositioned(newField.Identifier),
+			}
+		}
+
+		err := validator.checkField(newField, oldField)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (validator *ContractUpdateValidator) checkField(
+	newField *ast.FieldDeclaration,
+	oldField *ast.FieldDeclaration,
```

Comment:
All other check function have parameters ordered old-new instead of new-old. Maybe swap